lib/mix/lib/mix/sync/pubsub.ex

```diff
@@ -0,0 +1,292 @@
+defmodule Mix.Sync.PubSub do
+  @moduledoc false
+
+  # Pub/sub implementation working across multiple OS processes.
+  #
+  # The pub/sub is implemented using TCP sockets.
+  #
+  # Every subscriber opens a TCP socket and creates a port_P file.
+  # A publisher lists the directory, connects to each port and sends
+  # the message payload. When subscriber processes terminate, the
+  # files are left behind, so whenever a publisher fails to connect
+  # to one of the ports, it removes the corresponding file.
+  #
+  # We use a single socket for all the subscriptions. We create the
+  # socket lazily on the first subscription. Since the socket is
+  # shared, we include the hashed key in the broadcast payload, so
+  # that we can determine which subscribers to notify.
+
+  use GenServer
+
+  @opaque subscription :: %{socket: :gen_tcp.socket()}
+
+  @loopback {127, 0, 0, 1}
+  @listen_opts [:binary, ip: @loopback, packet: :raw, nodelay: true, backlog: 128, active: false]
+  @connect_opts [:binary, packet: :raw, nodelay: true, active: false]
+
+  @tag_data "mixpubsub"
+  @tag_data_size byte_size(@tag_data)
+
+  @name __MODULE__
+
+  @doc false
+  def start_link(_opts) do
+    GenServer.start_link(__MODULE__, {}, name: @name)
+  end
+
+  @doc """
+  Subscribes the caller to messages for `key`.
+
+  The messages are delivered to the caller's message queue.
+  """
+  @spec subscribe(String.t()) :: :ok
+  def subscribe(key) do
+    case GenServer.call(@name, {:subscribe, self(), key}, :infinity) do
+      :ok -> :ok
+      {:error, message} -> Mix.raise(message)
+    end
+  end
+
+  @doc """
+  Sends `message` to all processes subscribed to `key`.
+  """
+  @spec broadcast(String.t(), term() | (-> term())) :: :ok
+  def broadcast(key, message) do
+    hash = hash(key)
+    path = path(hash)
+
+    case File.ls(path) do
+      {:ok, []} ->
+        :ok
+
+      {:ok, names} ->
+        message =
+          case message do
+            lazy_message when is_function(lazy_message, 0) -> lazy_message.()
+            message -> message
```

Comment:
Do we want to always pass a function, just in case?

Reply:
Your call, we may use the pub/sub for other cases where there's nothing to compute, so forcing a function could be weird :)